argo/providers/vai/provider.go

```diff
@@ -249,60 +219,126 @@ func (vaip VAIProvider) DeleteRun(_ context.Context, _ VAIProviderConfig, _ stri
 	return nil
 }
 
+func buildPipelineJob(ctx context.Context, providerConfig VAIProviderConfig, runScheduleDefinition RunScheduleDefinition, mutatePipelineJob func(context.Context, VAIProviderConfig, *aiplatformpb.PipelineJob) error) (*aiplatformpb.PipelineJob, error) {
+	parameters := make(map[string]*aiplatformpb.Value, len(runScheduleDefinition.RuntimeParameters))
+	for name, value := range runScheduleDefinition.RuntimeParameters {
+		parameters[name] = &aiplatformpb.Value{
+			Value: &aiplatformpb.Value_StringValue{
+				StringValue: value,
+			},
+		}
+	}
+
+	// TODO migrate to non deprecated `ParameterValues` rather than `Parameters` below
+	pipelineJob := &aiplatformpb.PipelineJob{
+		Labels:         runLabelsFromSchedule(runScheduleDefinition),
+		TemplateUri:    providerConfig.pipelineUri(runScheduleDefinition.PipelineName.Name, runScheduleDefinition.PipelineVersion),
+		ServiceAccount: providerConfig.VaiJobServiceAccount,
+		RuntimeConfig: &aiplatformpb.PipelineJob_RuntimeConfig{
+			Parameters: parameters,
+		},
+	}
+
+	if err := mutatePipelineJob(ctx, providerConfig, pipelineJob); err != nil {
+		return nil, err
+	}
+
+	return pipelineJob, nil
+}
+
+func (vaip VAIProvider) buildCreatePipelineJobRequest(ctx context.Context, providerConfig VAIProviderConfig, runScheduleDefinition RunScheduleDefinition) (*aiplatformpb.Schedule_CreatePipelineJobRequest, error) {
+	pipelineJob, err := buildPipelineJob(ctx, providerConfig, runScheduleDefinition, vaip.specFromTemplateUri)
+	if err != nil {
+		return nil, err
+	}
+
+	return &aiplatformpb.Schedule_CreatePipelineJobRequest{
+		CreatePipelineJobRequest: &aiplatformpb.CreatePipelineJobRequest{
+			Parent:      providerConfig.parent(),
+			PipelineJob: pipelineJob,
+		},
+	}, nil
+}
+
+func (vaip VAIProvider) buildVaiSchedule(ctx context.Context, providerConfig VAIProviderConfig, runScheduleDefinition RunScheduleDefinition, buildPipelineJob func(context.Context, VAIProviderConfig, RunScheduleDefinition) (*aiplatformpb.Schedule_CreatePipelineJobRequest, error)) (*aiplatformpb.Schedule, error) {
+	cron, err := ParseCron(runScheduleDefinition.Schedule)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := buildPipelineJob(ctx, providerConfig, runScheduleDefinition)
```

Comment:
I See what you did here for testability 👍 
The more GO~diotic~diomatic way to this would be to use mixins I believe: https://stackoverflow.com/a/44707721

Reply:
Will get my go cap on

Reply:
Done